pyccel/parser/base.py

```diff
@@ -500,17 +503,118 @@ def insert_import(self, expr):
 
-        # TODO improve spacing
+          Parameters
+          ----------
+          filename: str
+            output file name. if not given `name.pyccel` will be used and placed
+            in the Pyccel directory ($HOME/.pyccel)
+        """
+        import pickle
+        import hashlib
+
+        if not filename:
+            if not self.filename:
+                raise ValueError('Expecting a filename to load the ast')
+
+            path , name  = os.path.split(self.filename)
+
+            if not name.split('.')[-1] == 'pyh':
+                return
+            else:
+                name = name[:-4]
 
-        print ('------- namespace -------')
-        for (k, v) in self.namespace.items():
-            print ('{var} \t :: \t {dtype}'.format(var=k, dtype=type(v)))
-        print ('-------------------------')
+            name     = '{}.pyccel'.format(name)
+            filename = path+ '/' + name
+        # check extension
 
-    def _visit(self, expr, **settings):
-        raise NotImplementedError('Must be implemented by the extension')
+        if not filename.split(""".""")[-1] == 'pyccel':
+            raise ValueError('Expecting a .pyccel extension')
+
+#        print('>>> home = ', os.environ['HOME'])
+        # ...
+
+        # we are only exporting the AST.
+        try:
+            code = self.code.encode('utf-8')
+            hs   = hashlib.md5(code)
+            f    = open(filename, 'wb')
+            pickle.dump((hs.hexdigest(), __version__, self), f, pickle.HIGHEST_PROTOCOL)
+            f.close()
+        except:
+            return
+
+    # TODO shall we need to load the Parser too?
+
+    def load(self, filename=None):
+        """
+        Load the current ast using Pickle.
+
+          Parameters
+          ----------
+          filename: str
+            output file name. if not given `name.pyccel` will be used and placed
+            in the Pyccel directory ($HOME/.pyccel)
+        """
+
+        # ...
+        import pickle
+
+        if not filename:
+            if not self.filename:
+                raise ValueError('Expecting a filename to load the ast')
+
+            path , name = os.path.split(self.filename)
+
+            if not name.split('.')[-1] == 'pyh':
+                return
+            else:
+                name = name[:-4]
+
+            name     = '{}.pyccel'.format(name)
+            filename = path+ '/' + name
+        # check extension
+
+        if not filename.split(""".""")[-1] == 'pyccel':
+            raise ValueError('Expecting a .pyccel extension')
+
+#        print('>>> home = ', os.environ['HOME'])
+        # ...
+        try:
+            f = open(filename, 'rb')
+            hs, version, parser = pickle.load(f)
+            f.close()
+        except:
+            return
+
+        import hashlib
+        code = self.code.encode('utf-8')
+        if hashlib.md5(code).hexdigest() == hs and __version__ == version:
+            self.copy(parser)
```

Comment:
Is there any way we can test this? Maybe a test with 2 versions of a `.pyh` and a `cp` command to create the `.pyh` file that we are working with here?

Reply:
I agree, we need to test this mechanism somehow. And we should not take for granted that it works on all platforms.

Reply:
it can be tested if we change the version so I'm not sure how can we change the version

Reply:
the user needs to update the repository and then it will be tested

Reply:
So we only recreate the file if we update the pyccel version? Is that necessary? I thought the `.pyh` files let you interact with any fortran code. So in theory a user could create their own. Given that fact I'm not sure it makes sense to link this to the pyccel version

Reply:
It's necessary if we change something on the way we treat the header files, imagine we change the syntax in the new version or we change the something the outcome of the parser in the new version, so there is no way for us the be sure

Reply:
this part of code deals with `.pyccel` files and not `-pyh` files

Reply:
Yes, but the `.pyccel` files are generated from the `.pyh` files aren't they?

My point was that I might want to update the `.pyccel` file between version changes, not that we shouldn't update it when we update the version

Reply:
yes this make sure that we update them when we change the version or the pyh file

Reply:
In that case we should be able to test it as I suggested. You create a file containing:
- test_file_v1.pyh
- test_file_v2.pyh
- test_script.py

In test_script.py you do:
```
cp test_file_v1.pyh test_file.pyh
pyccel test_file.pyh
# Test output expected from v1
cp test_file_v2.pyh test_file.pyh
pyccel test_file.pyh
# Test output expected from v2
```

Reply:
done